File: ozys-app/src-tauri/src/ozys_device.rs
use anyhow::Result;
use async_trait::async_trait;
use serde::Serialize;

#[derive(Serialize, Clone)]
pub struct OzysDeviceInfo {
    pub name: String,
    pub id: String,
    pub model: String,
    pub channels: Vec<OZYSChannelState>,
}

#[derive(Serialize, Clone, PartialEq, Eq)]
#[serde(tag = "state")]
pub enum OZYSChannelState {
    Disconnected,
    Connected {
        enabled: bool,
        name: String,
        id: String,
    },
}

/// The frontend polls this at 10Hz, contains the readings and
/// FFT of the measurements from the last 100ms
#[derive(Serialize, Clone)]
pub struct OzysChannelRealtimeData {
    /// Absolute reading values
    /// Unit is 1, e.g. 0.01 = 1% of length change
    /// 10 readings long, interval between readings is 10ms
    pub readings: Vec<f32>,

    /// Standard deviation of all the samples inside the 10ms interval
    /// 10 values long, one value for each reading
    pub reading_noises: Vec<f32>,

    /// FFT for frequencies below 2kHz, with 10Hz resolution
    /// 200 values long
    /// e.g. fft_0_to_2k[0] is the power of the 0-10Hz band in the last 100ms
    ///      fft_0_to_2k[1] is the power of the 10-20Hz band in the last 100ms
    pub fft_0_to_2k: Vec<f32>,

    /// FFT for frequencies between 2kHz and 20kHz, with 50Hz resolution
    /// 360 values long
    /// e.g. fft_2k_to_20k[0] is the power of the 2k-2.05k band in the last 100ms
    pub fft_2k_to_20k: Vec<f32>,
}

#[async_trait]
pub trait OzysDevice {
    fn get_device_info(&self) -> OzysDeviceInfo;
    async fn rename_device(&mut self, new_name: String) -> Result<()>;
    async fn rename_channel(&mut self, channel_index: usize, new_name: String) -> Result<()>;
    async fn control_channel(&mut self, channel_index: usize, enabled: bool) -> Result<()>;
    async fn control_recording(&mut self, record: bool) -> Result<()>;

    /// The frontend should poll this at 10Hz
    /// Returns Ok(None) if polling is too frequent
    async fn poll_realtime_data(&mut self) -> Result<Option<Vec<Option<OzysChannelRealtimeData>>>>;
}

File: ozys-app/src/device/OzysDevice.ts
import { action, makeAutoObservable, makeObservable, observable } from 'mobx'

export type OzysChannelState =
  | {
      connected: false
    }
  | {
      connected: true
      enabled: boolean
      name: string
      id: string
    }

export type OzysDeviceInfo = {
  name: string
  id: string
  model: string
  isRecording: boolean
  channels: OzysChannelState[]
}

// The device sends this data to the app every 10ms
// The readings list contains 20 samples during the last 10ms
// Sampling rate is 2kHz
export type OzysChannelRealtimeReadings = {
  timestamp: number

  // Absolute reading values
  // Unit is 1, e.g. 0.01 = 1% of length change
  // 20 readings long, interval between readings is 0.5ms
  readings: Float32Array

  // Standard deviation of all the samples inside the 10ms interval
  // 20 values long, one value for each reading
  readingNoises: Float32Array
}

// The device sends this data to the app every 100ms
// The data contains FFT during the last 100ms
export type OzysChannelRealtimeFft = {
  timestamp: number

  // FFT for frequencies below 2kHz, with 10Hz resolution
  // 200 values long
  // e.g. fft_0_to_2k[0] is the power of the 0-10Hz band in the last 100ms
  //      fft_0_to_2k[1] is the power of the 10-20Hz band in the last 100ms
  fft0To2k: Float32Array

  // FFT for frequencies between 2kHz and 20kHz, with 50Hz resolution
  // 360 values long
  // e.g. fft_2k_to_20k[0] is the power of the 2k-2.05k band in the last 100ms
  fft2kTo20k: Float32Array
}

export abstract class OzysDevice {
  protected readingCallbacks: Array<
    (channelId: string, data: OzysChannelRealtimeReadings) => void
  > = []
  protected fftCallbacks: Array<
    (channelId: string, data: OzysChannelRealtimeFft) => void
  > = []

  constructor(public deviceInfo: OzysDeviceInfo) {
    makeObservable(this, {
      deviceInfo: observable,
      renameDevice: action,
      renameChannel: action,
      controlChannel: action,
      controlRecording: action,
    })
  }

  async renameDevice(name: string) {
    this.deviceInfo.name = name
  }

  async renameChannel(channelId: string, name: string) {
    for (const channel of this.deviceInfo.channels) {
      if (channel.connected && channel.id === channelId) {
        channel.name = name
        return
      }
    }
  }

  async controlChannel(channelId: string, enabled: boolean) {
    for (const channel of this.deviceInfo.channels) {
      if (channel.connected && channel.id === channelId) {
        channel.enabled = enabled
        return
      }
    }
  }

  async controlRecording(enabled: boolean) {
    this.deviceInfo.isRecording = enabled
  }

  onRealtimeReadings(
    callback: (channelId: string, data: OzysChannelRealtimeReadings) => void,
  ): () => void {
    this.readingCallbacks.push(callback)
    return () => {
      this.readingCallbacks = this.readingCallbacks.filter(
        (cb) => cb !== callback,
      )
    }
  }

  onRealtimeFft(
    callback: (channelId: string, data: OzysChannelRealtimeFft) => void,
  ): () => void {
    this.fftCallbacks.push(callback)
    return () => {
      this.fftCallbacks = this.fftCallbacks.filter((cb) => cb !== callback)
    }
  }

  abstract disconnect(): void
}
